From 41ac2d3e9878379bf1a5a17626d489acae39bdba Mon Sep 17 00:00:00 2001
From: Anca-Jeanina Floarea <anca.floarea@freescale.com>
Date: Thu, 29 Nov 2012 23:01:55 +0000
Subject: [PATCH 420/518] dpa_ipsec: Change flow id location based on
 DPAA_VERSION

For FMAN v2 devices, the flow id needs to be written in ContextB,
while for FMAN v3 devices it is written in ContextA

Signed-off-by: Anca Jeanina FLOAREA <anca.floarea@freescale.com>
[Grabbed from the branch, LINUX_IR5.2.0, of
https://git.freescale.com/git-private/cgit.cgi/ppc/alu-b4860/linux.git.]
Signed-off-by: Tiejun Chen <tiejun.chen@windriver.com>
---
 drivers/staging/fsl_dpa_offload/dpa_ipsec.c |   14 ++++++++++++--
 drivers/staging/fsl_dpa_offload/dpa_ipsec.h |    2 ++
 2 files changed, 14 insertions(+), 2 deletions(-)

diff --git a/drivers/staging/fsl_dpa_offload/dpa_ipsec.c b/drivers/staging/fsl_dpa_offload/dpa_ipsec.c
index 4519d84..e3f452c 100644
--- a/drivers/staging/fsl_dpa_offload/dpa_ipsec.c
+++ b/drivers/staging/fsl_dpa_offload/dpa_ipsec.c
@@ -2133,7 +2133,7 @@ static int create_sec_frame_queue(uint32_t fq_id, uint16_t channel,
 		if (sp_op) {
 			void *fm = NULL;
 
-			/* ugly hack to get FMan handle from the PCD handle*/
+			/* get FMan handle from the PCD handle*/
 			fm = ((t_FmPcd *)fm_pcd)->h_Fm;
 
 			error = FM_GetSpecialOperationCoding(fm, sp_op,
@@ -2143,10 +2143,20 @@ static int create_sec_frame_queue(uint32_t fq_id, uint16_t channel,
 				pr_err("Could not retrieve special op code");
 				goto create_sec_fq_err;
 			}
+			/* the opcode is a 4-bit value */
+			sp_op_code &= NIA_OPCODE_MASK;
 		}
+#if (DPAA_VERSION == 10)
+		/* FMAN v2 devices: opcode and flow id are stored in contextB */
 		FM_CONTEXTA_SET_OVERRIDE(&fq_opts.fqd.context_a, true);
 		FM_CONTEXTB_SET_FQID(&(fq_opts.fqd.context_b), ctxB |
-				     (sp_op_code << 20));
+					(sp_op_code << 20));
+#elif (DPAA_VERSION == 11)
+		/* FMAN v3 devices: opcode and flow id are stored in contextA */
+		FM_CONTEXTA_SET_A1_VALID(&fq_opts.fqd.context_a, true);
+		FM_CONTEXTA_SET_A1(&fq_opts.fqd.context_a,
+				((ctxB << 4) | sp_op_code));
+#endif
 	}
 
 	fq_opts.fqd.dest.wq = wq_id;
diff --git a/drivers/staging/fsl_dpa_offload/dpa_ipsec.h b/drivers/staging/fsl_dpa_offload/dpa_ipsec.h
index 0c2f713..0fd7264 100644
--- a/drivers/staging/fsl_dpa_offload/dpa_ipsec.h
+++ b/drivers/staging/fsl_dpa_offload/dpa_ipsec.h
@@ -191,6 +191,8 @@
 /* The maximum length (in bytes) for the CAAM extra commands */
 #define MAX_EXTRA_DESC_COMMANDS		(64 * sizeof(uint32_t))
 
+#define NIA_OPCODE_MASK		0x0F
+
 /*
  * calculate the size of the fields in the policy params structure that
  * are used to identify a policy
-- 
1.7.5.4

